### ISSUE TITLE
how to manage custom functions?

### ISSUE THREAD

**Issue by JoernT** [FontoXML/fontoxpath#256]
This is kind of a follow-up to my question in issue #108 

First of all, thanks again for your answer. I just returned to actively work on my project and started to look into custom function implementation after my first proof-of-concept worked out nicely thanks to FontoXPath. 

As i'll need quite a bunch of functions that are specific to XForms the question arises how to manage those function from a coding perspective. You write i should just add my declarations to builtInFunctions.ts. But that means i'll always have to maintain my copy of fontXpath myself right?

Is there any mechanism to load functions from an external module or are there plans for that?

It's not a KO criterium but certainly would make life easier for me.

**Comment by DrRataplan** [comment]
Hi,

The main extension point for functions is [`registerCustomXPathFunction`](https://github.com/FontoXML/fontoxpath/blob/master/src/registerCustomXPathFunction.ts). This allows you to implement an external function in JavaScript and can be used for interopts to other systems.

If you want to implement a function that does not need interopts, I'd suggest taking a look at [`registerXQueryModule`](https://github.com/FontoXML/fontoxpath/blob/master/src/registerXQueryModule.ts), which allows you to input your own modules. These modules can then be accessed using the normal XQuery constructs, or by using the [`moduleImports`](https://github.com/FontoXML/fontoxpath/blob/master/src/evaluateXPath.ts#L30) in a place where you want to use them without explicitly importing them.

Hope this helps you,

Martin

**Comment by JoernT** [comment]
hm, 

i call registerCustomXPathFunction in my code but keep getting
`Uncaught Error: XPST0017: Function instance with arity of 1 not registered. Did you mean "Q{}instance (xs:string?)"?`

(using empty string for namespace but also tried with prefix)

Calling it like this somewhere in my init code. 

```
registerCustomXPathFunction(
            { namespaceURI: '', localName: 'instance' },
            ['xs:string?'],
            'node()',
            (dynamicContext, string) => {
                console.log('fnInstance: ', dynamicContext);
                console.log('fnInstance: ', dynamicContext.domFacade);
                return string === null || string === 'test';
            }
        );
```
Copied that stuff from the tests.

Guess there's something wrong with my context.

**Comment by DrRataplan** [comment]
Ah, that is most likely because the null namespace is not allowed to have functions. We should throw an error for that.

I've quickly mashed together an example:https://runkit.com/drrataplan/5efb4e09788e1a001a6d2a47

I'll also include a couple of examples in the ready, to clarify what is what a bit better.

**Comment by JoernT** [comment]
Yes, in fact the functions should be in the xforms namespace but be used without a prefix which always makes it a bit hacky. 

No idea how to achieve that otherwise so was a dirty try ;)

**Comment by DrRataplan** [comment]
Setting the default function namespace could also work? just prefix every query you have with `'declare default function namespace "the xforms namespace uri";'` Unfortunately, we do not have API yet to override the default function namespace without manipulating the XQuery string, the namespaceresolver property in evaluateXPath is only overriding the default element namespace.

**Comment by JoernT** [comment]
getting a bit further - at least i can evaluate the function with the QName syntax. 

But there are problems remaining. 

I try to outline my use-case a bit more... no obligations to follow me here as it's probably quite special but i'm of course thankful for any comments.

following code snippet illustrates the goal. 

```
    <xf-form>
        <xf-model id="model1">
            <xf-instance id="default">
                <data>
                    <greeting>Hello World!</greeting>
                </data>
            </xf-instance>

            <xf-bind ref="greeting" required="1 = 1"></xf-bind>

            <xf-instance id="second">
                <data>
                    <outro>GoodBye</outro>
                </data>
            </xf-instance>

            <xf-bind ref="instance('second')//outro"></xf-bind>

        </xf-model>
    </xf-form>
```

The above tags are HTML5 web components in the body of an html document.

Every xf-instance element wraps an standalone XML document that is created from the inline content and has an `id`. When the model initializes the xf-bind/@ref expressions (plain XPath) are evaluated and resolve to nodes in the instances. 

First bind resolves to the value 'Hello World!'. Second bind uses the `instance` function to resolve to the second instance document and resolve to 'GoodBye'. 

2 Problems here:
* first i can't manage to use the above syntax `instance('second')` as i cannot inject a default namespace for the xforms functions 
* resolution fails when combining with locationPath

the function look like this:
```
 registerCustomXPathFunction(
            { namespaceURI: 'xf', localName: 'instance' },
            ['xs:string'],
            'node()',
            (dynamicContext, string) => {
                console.log('fnInstance dynamicContext: ', dynamicContext);
                console.log('fnInstance string: ', string);

                const instance = this.querySelector('xf-instance[id=' + string + ']');
                return instance.getDefaultContext();
            }
        );
```
The callback fetches the `xf-instance` element. This provides a getDefaultContext() which returns an Element
```
        const instanceData = new DOMParser().parseFromString(this.innerHTML,'text/xml');
        return this.instanceData.firstElementChild;
```

Just the function as expression works:
```
        fx.evaluateXPath('Q{xf}instance("second")',this.model,null,{});
```


and this is finally the failing call:
```
        const result = fx.evaluateXPath('Q{xf}instance("second")//outro',this.model,null,{});
```


Thanks for listening.

**Comment by DrRataplan** [comment]
For the default function namespace thing, I'd be ok with having additional
API that sets that. Same as moduleImports or namespace resolver.

Quick question, I'd assume all of the fn functions need the fn prefix in
xforms, right? Such as true(), false(), empty()?

On the second point, at first sight I see nothing wrong. I'll have to try
that out myself. Can you describe what is awry? Is it an error, empty
results, or otherwise? Maybe place a few fn:trace calls here and there?

**Comment by JoernT** [comment]
On Tue, Jun 30, 2020 at 7:54 PM Martin Middel <notifications@github.com>
wrote:

> For the default function namespace thing, I'd be ok with having additional
> API that sets that. Same as moduleImports or namespace resolver.
>
> Quick question, I'd assume all of the fn functions need the fn prefix in
> xforms, right? Such as true(), false(), empty()?
>

You mean writing them as fnTrue() etc.? Or being the fn namespace?

In XForms you use all functions as 'usual' writing true(), false() etc.

All common XPath functions are in the Static Context in XForms with the
namespaces being in scope.

Cite from XForms function spec:
"http://www.w3.org/2002/xforms-functions

   - This namespace exposes all XForms Function Library functions as well
   as all standard XPath functions.
   - The XPath static context's default function namespace must be set to
   this URI."

As the XForms spec is a bit behind time it still relies on XPath 2 but i'll
ignore that ;) I'm taking freedom from the spec this time - in our original
implementation in Java we took things very strict and reaches 99.7% spec
conformance. But that took about 10 years ;) This time i take it a bit
easier and let things evolve along the requirements of users. Often a
subset is already much more than all other form frameworks around.

Just in case you're interested - the function spec is here -
https://www.w3.org/community/xformsusers/wiki/XPath_Expressions_Module

I'll certainly not implement all the functions at once - again, following
what we hit in projects and implement things as the get required.


>
> On the second point, at first sight I see nothing wrong. I'll have to try
> that out myself. Can you describe what is awry? Is it an error, empty
> results, or otherwise? Maybe place a few fn:trace calls here and there?
>

Look at this:

let result = fx.evaluateXPath('Q{xf}instance("second")',this.model,null,{});
console.log('????? result ',result);
result = fx.evaluateXPath('Q{xf}instance("second")//outro',this.model,null,{});
console.log('????? result ',result);

the first evaluation correctly prints the XML of instance 'second' to
console.
the second evaluation returns an empty array.

**Comment by JoernT** [comment]
my current assumption is that it has to do with the fact that the evaluation context is the xf-model element but instance('outro') returns a different document and cannot resolve any pathes that dig into that?

However this is contradicted by the fact that 
`evaluateXPath('Q{xf}instance("second")/*',this.model,null,{debug:true});`

returns 
```
<data>
    <outro>GoodBye</outro>
</data>
```

But all attempts to use something different than '*' like '//' or plain locationpath does not match anything.

Hmm...

**Comment by DrRataplan** [comment]
After a discussion on another medium, we concluded that the issue is namespace related. The default namespace on the `this.model` variable is not the null namespace. The default resolver is used here, which is incorrect; a specific resolver should be written that resolves prefixes based on the xforms host document.

**Comment by JoernT** [comment]
ok, the evaluation issue is gone now by using the right contextnode.

However there's stil the issue how to implement custom functions that can be used along with the standard XPath functions without using a prefix like e.g.

`instance("foo")'`
`boolean-from-string`
etc. 

So i guess the functions must share the default fn namespace OR establish a new namespace that exposes all standard functions AND the XForms extension functions in one namespace. See my cite from the XForms Spec. above.

**Comment by DrRataplan** [comment]
You can do one of two things here: prefix every query with `declare default function namespace="HTTP://something-something-xforms"` to set the default function namespace to the 'xforms' namespace and re-expose all of the 'fn' functions to the 'xforms' namespace. See spec for the default function namespace: https://www.w3.org/TR/xquery-31/#doc-xquery31-DefaultNamespaceDecl

The other way (which I'd prefer for long term) is to expose a way to set this namespace URI using the JavaScript API. I'm open for a PR here.
